### PR TITLE
 Add jit tests for `alias`, `always_inline`, `noinline`, `used` and `weak` attributes

### PR DESCRIPTION
**Pull request by GuillaumeGomez**

### CODE REVIEW

File: gcc/testsuite/jit.dg/test-always_inline-attribute.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
+/* { dg-do compile { target x86_64-*-* } } */
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "libgccjit.h"
+
+/* We don't want set_options() in harness.h to set -O3 to see that the pure
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-always_inline-attribute.c:8]:
```suggestion
/* We don't want set_options() in harness.h to set -O3 to see that the always_inline
```

File: gcc/testsuite/jit.dg/test-always_inline-attribute.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    GCC_JIT_BINARY_OP_PLUS,
+    gcc_jit_context_new_call (ctxt, NULL, not_removed_func, 1, &null));
+
+  /* return x; */
+  gcc_jit_block_end_with_return (foo_block, NULL, gcc_jit_lvalue_as_rvalue(x));
+}
+
+/* { dg-final { jit-verify-output-file-was-created "" } } */
+/* Check that the "removed" function was inlined, but not the others */
+/* { dg-final { jit-verify-assembler-output-not ".type	removed, @function" } } */
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-always_inline-attribute.c:153]: I don't like that this is format-sensitive.
Would it work with:
```suggestion
/* { dg-final { jit-verify-assembler-output-not "removed" } } */
```
or:
```suggestion
/* { dg-final { jit-verify-assembler-output-not "removed, @function" } } */
```

File: gcc/testsuite/jit.dg/test-noinline-attribute.c [hunk 3]
```diff
@@ -0,0 +1,114 @@
+/* { dg-do compile { target x86_64-*-* } } */
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "libgccjit.h"
+
+/* We don't want set_options() in harness.h to set -O3 to see that the pure
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-noinline-attribute.c:8]:
```suggestion
/* We don't want set_options() in harness.h to set -O3 to see that the noinline
```

File: gcc/testsuite/jit.dg/test-noinline-attribute.c [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+#define TEST_ESCHEWS_SET_OPTIONS
+static void set_options (gcc_jit_context *ctxt, const char *argv0)
+{
+  // Set "-O2".
+  gcc_jit_context_set_int_option(ctxt, GCC_JIT_INT_OPTION_OPTIMIZATION_LEVEL, 2);
+}
+
+#define TEST_COMPILING_TO_FILE
+#define OUTPUT_KIND      GCC_JIT_OUTPUT_KIND_ASSEMBLER
+#define OUTPUT_FILENAME  "output-of-test-alwasy_inline-attribute.c.s"
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-noinline-attribute.c:19]:
```suggestion
#define OUTPUT_FILENAME  "output-of-test-noinline-attribute.c.s"
```

File: gcc/testsuite/jit.dg/test-noinline-attribute.c [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    GCC_JIT_BINARY_OP_PLUS,
+    gcc_jit_context_new_call (ctxt, NULL, not_removed_func, 0, NULL));
+
+  /* return x; */
+  gcc_jit_block_end_with_return (foo_block, NULL, gcc_jit_lvalue_as_rvalue(x));
+}
+
+/* { dg-final { jit-verify-output-file-was-created "" } } */
+/* Check that the "removed" function was inlined, but not the others */
+/* { dg-final { jit-verify-assembler-output-not ".type	removed, @function" } } */
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-noinline-attribute.c:112]: Same.

File: gcc/testsuite/jit.dg/test-used-attribute.c [hunk 6]
```diff
@@ -0,0 +1,114 @@
+/* { dg-do compile { target x86_64-*-* } } */
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "libgccjit.h"
+
+/* We don't want set_options() in harness.h to set -O3 to see that the pure
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-used-attribute.c:8]:
```suggestion
/* We don't want set_options() in harness.h to set -O3 to see that the used
```

File: gcc/testsuite/jit.dg/test-used-attribute.c [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+   attribute affects the optimizations. */
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-used-attribute.c:9]: Maybe something else than "optimizations" here?

File: gcc/testsuite/jit.dg/test-used-attribute.c [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+#define TEST_ESCHEWS_SET_OPTIONS
+static void set_options (gcc_jit_context *ctxt, const char *argv0)
+{
+  // Set "-O2".
+  gcc_jit_context_set_int_option(ctxt, GCC_JIT_INT_OPTION_OPTIMIZATION_LEVEL, 2);
+}
+
+#define TEST_COMPILING_TO_FILE
+#define OUTPUT_KIND      GCC_JIT_OUTPUT_KIND_ASSEMBLER
+#define OUTPUT_FILENAME  "output-of-test-alwasy_inline-attribute.c.s"
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-used-attribute.c:19]:
```suggestion
#define OUTPUT_FILENAME  "output-of-test-used-attribute.c.s"
```

File: gcc/testsuite/jit.dg/test-used-attribute.c [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    GCC_JIT_BINARY_OP_PLUS,
+    gcc_jit_context_new_call (ctxt, NULL, not_removed_func, 0, NULL));
+
+  /* return x; */
+  gcc_jit_block_end_with_return (foo_block, NULL, gcc_jit_lvalue_as_rvalue(x));
+}
+
+/* { dg-final { jit-verify-output-file-was-created "" } } */
+/* Check that the "removed" function was inlined, but not the others */
+/* { dg-final { jit-verify-assembler-output-not ".type	removed, @function" } } */
```
Comment (antoyo) [inline gcc/testsuite/jit.dg/test-used-attribute.c:112]: Same.

File: gcc/testsuite/jit.dg/test-always_inline-attribute.c (hunk 2 above)
Comment (GuillaumeGomez) [reply gcc/testsuite/jit.dg/test-always_inline-attribute.c:153]: From what I could gather, it's not format sensitive as it's always following this format syntax. The important thing tested here is to ensure that the `not_removed` function is present (with `@function`) whereas `removed` is not.

File: gcc/testsuite/jit.dg/test-noinline-attribute.c (hunk 5 above)
Comment (GuillaumeGomez) [reply gcc/testsuite/jit.dg/test-noinline-attribute.c:112]: Same. :)

File: gcc/testsuite/jit.dg/test-used-attribute.c (hunk 7 above)
Comment (GuillaumeGomez) [reply gcc/testsuite/jit.dg/test-used-attribute.c:9]: Let's remove the comment...

File: gcc/testsuite/jit.dg/test-used-attribute.c (hunk 9 above)
Comment (GuillaumeGomez) [reply gcc/testsuite/jit.dg/test-used-attribute.c:112]: Same. :)

**Pull request merged**